File: experiments/runs/2026-08-10_15-25-37_claim-office-example-mapping_v3-basic-tdd_opus-5-no-thinking-2/src/policy.ts
import { insuranceValueOf, priceOf } from './catalog.js';
import { ClaimOfficeError, type ClaimResult, type Damage, type Incident, type Item } from './types.js';

const DEDUCTIBLE = 100;
const CAP_FACTOR = 2;
const HIGH_ENCHANTMENT_LEVEL = 8;
const HIGH_ENCHANTMENT_REIMBURSEMENT = 0.5;

/**
 * Reimbursement rate for a damaged item. Damage is reimbursed in full unless
 * the item is highly enchanted; that clause takes precedence over the full
 * reimbursement dragon material would otherwise guarantee.
 */
function reimbursementRate(item: Item): number {
  const highlyEnchanted = (item.enchantment ?? 0) >= HIGH_ENCHANTMENT_LEVEL;
  return highlyEnchanted ? HIGH_ENCHANTMENT_REIMBURSEMENT : 1;
}

/** An insurance policy created by a quote, tracking its remaining cap. */
export class Policy {
  readonly insuranceSum: number;
  remainingCap: number;

  constructor(private readonly items: Item[]) {
    this.insuranceSum = items.reduce((sum, item) => sum + insuranceValueOf(item), 0);
    this.remainingCap = this.insuranceSum * CAP_FACTOR;
  }

  /**
   * Matches each damage to a distinct covered item of the same type, so that
   * a policy is never claimed against more often than it covers.
   */
  private matchDamages(damages: Damage[]): { damage: Damage; item: Item }[] {
    const available = [...this.items];
    return damages.map((damage) => {
      priceOf(damage.itemType);
      if (damage.amount < 0) {
        throw new ClaimOfficeError(
          `damage amount must not be negative: ${damage.amount} for ${damage.itemType}`,
        );
      }
      const index = available.findIndex((item) => item.type === damage.itemType);
      if (index === -1) {
        throw new ClaimOfficeError(`item not covered by the policy: ${damage.itemType}`);
      }
      const [item] = available.splice(index, 1);
      return { damage, item };
    });
  }

  /** Processes a damage report, consuming part of the policy's cap. */
  claim(incident: Incident): ClaimResult {
    const matched = this.matchDamages(incident.damages);

    const desired = matched.reduce((sum, { damage, item }) => {
      const reimbursed = damage.amount * reimbursementRate(item);
      return sum + Math.max(0, reimbursed - DEDUCTIBLE);
    }, 0);

    const payout = Math.floor(Math.min(desired, this.remainingCap));
    this.remainingCap -= payout;

    return { payout, remainingCap: this.remainingCap };
  }
}

File: experiments/runs/2026-09-22_01-57-46_claim-office-python-example-mapping_baseline-inline-tdd-v1-cc_opus-5-no-thinking/src/policy.py
"""A policy: the items it covers and the cap it has left."""

from collections import Counter

import claim as claim_rules
import pricelist
from rounding import round_in_favour_of_customer

CAP_FACTOR = 2


class ClaimError(Exception):
    """Raised when a claim cannot be processed against this policy."""


class ClaimResult:
    """The outcome of one claim against a policy."""

    def __init__(self, payout, remaining_cap):
        self.payout = payout
        self.remaining_cap = remaining_cap


class Policy:
    """Items insured for one customer, tracking the cap already used up."""

    def __init__(self, customer, items):
        self.customer = customer
        self.items = items
        self.insurance_sum = sum(pricelist.insurance_value(i["type"]) for i in items)
        self.remaining_cap = self.cap

    @property
    def cap(self):
        return CAP_FACTOR * self.insurance_sum

    def claim(self, incident):
        """Process one incident, reducing the remaining cap by the payout."""
        damages = incident["damages"]
        self._reject_negative_amounts(damages)
        self._reject_damages_not_covered(damages)
        desired = sum(
            claim_rules.reimbursement(self._item_for(d["itemType"]), d["amount"])
            for d in damages
        )
        payout = round_in_favour_of_customer(min(desired, self.remaining_cap))
        self.remaining_cap -= payout
        return ClaimResult(payout, self.remaining_cap)

    @staticmethod
    def _reject_negative_amounts(damages):
        for d in damages:
            if d["amount"] < 0:
                raise ClaimError(f"negative damage amount: {d['amount']}")

    def _item_for(self, item_type):
        return next(i for i in self.items if i["type"] == item_type)

    def _reject_damages_not_covered(self, damages):
        """Each damage entry needs an insured item of its own to claim against."""
        insured = Counter(i["type"] for i in self.items)
        claimed = Counter(d["itemType"] for d in damages)
        for item_type, count in claimed.items():
            if count > insured[item_type]:
                raise ClaimError(f"not insured often enough: {item_type}")
